### PR TITLE
POC Migrate Metaphysics to Hokusai / Kubernetes

### PR DESCRIPTION
**Pull request by izakp**
- Pin Node version to  8.4.0 as that is what is running on Heroku
- Setup Hokusai
- Remove `-w` flag from nf as it was causing [this bug](https://github.com/markmur/sails-react-webpack/issues/5) when it encountered "Wrapping display Output to NaN Columns"
- remove unneeded test env var `MEMCACHE_SERVERS`

I did this as a POC for building Metaphysics (and other Node apps) with Docker images, for then deploying them to Kubernetes using Hokusai.  I think it gives us a number of advantages over Heroku's build pipeline:

- We are able to configure the node runtime, npm and yarn versions and test against different node versions, as pinned by the first line `FROM node:8.4.0` in the `Dockerfile`.
- We are able to run the babel compiler inline in our build step and not reply on a `heroku-postbuild` task
- We are able to inspect and test the deploy artifact (Docker image) in an environment on parity with production

In terms of resource allocation and orchestration, Kubernetes gives us much more flexibility to scale horizontally quickly and easily, as well as fail over from memory leaks gracefully.

To set up Hokusai, follow the Requirements / Setup instructions [here](https://github.com/artsy/hokusai#requirements).

Our `hokusai config ...` org-specific options can be found [here](https://github.com/artsy/potential/blob/master/platform/Kubernetes.md#hokusai)

You can then run (in the project root) `hokusai dev` and Metaphysics will build and run (along with memcached) and be exposed on `localhost:5001`.  Delete the `MEMCACHE_SERVERS` line from your `.env` file, if present, as this is supplied by `./hokusai/development.yml` which points to memcached running in a separate container.  You can also run `hokusai test` to run the test suite inside a container.

This could greatly simplify all the different `.env.test` / `.env.staging` configuration as well as commands in `package.json`.  In terms of getting it launched on Kubernetes, we would have to:

- Update the `.circleci/config.yml` file to work with the Hokusai build and deploy pipeline, as in [APRb](https://github.com/artsy/aprb/blob/master/.circleci/config.yml)
- Create a memcached instance in Elasticache for Metaphysics
- Migrate the env vars set in Heroku over to Hokusai
- Switch DNS

### CODE REVIEW

File: hokusai/development.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    - MEMCACHED_SERVERS=metaphysics-memcached:11211
+    extends:
+      file: common.yml
+      service: metaphysics
+    ports:
+    - 5000:5000
+  metaphysics-memcached:
+    image: memcached
+    ports:
+    - 11211:11211
```
Comment (joeyAghion) [inline hokusai/development.yml]: Gotta love how simple this is.

**Comment by joeyAghion** [comment]
Very interesting to see this. Surprisingly simple!

If this is merged, presumably development will need to happen in-container going forward, right?

Also, just wondering if there's an opportunity to test this in a lightweight way. Like, could we support both K8 and Heroku for an interim period, and deploy staging to K8 for some realistic testing?

File: hokusai/production.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+          - containerPort: 5000
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    app: metaphysics
+    layer: application
+    component: web
+  name: metaphysics-web
```
Comment (ashkan18) [inline hokusai/production.yml]: Should we do`annotations` we did in other apps for forcing `https` stuff here too? 

https://github.com/artsy/lewitt-api/blob/master/hokusai/production.yml#L81

**Comment by izakp** [comment]
@joeyAghion development won't necessarily _have_ to happen in-container going forward but it would allow any dev to test the build locally via `hokusai dev` before pushing it to CI

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,27 @@
+FROM node:8.4.0
+
+# Install packages
+RUN npm install -g yarn
+
+# Set up deploy user and working directory
```
Comment (izakp) [inline Dockerfile]: Normally we could safely run as `root` in a container (as processes in a container are "sandboxed" and have no system-level root privileges by default but I encountered an error whereby node-foreman tried to drop sudo and failed: https://github.com/strongloop/node-foreman/issues/70

**Comment by orta** [comment]
I've just got pretty close to having this working for dev locally.

I have a blank computer, so no hokusai/python/AWS setup. We could do with making an artsy engineer user guide on  hokusai (on artsy/potential or artsy/meta so it's OSS) if this is going to be an encouraged way to dev services 👍 

Here are my notes from setting up as a first timer:

---

### Hokusai 

- Default python is 2.7 - should be fine
- Manually installed pip, had to use sudo
- Ran `pip install hokusai` - failed

- `brew install python` installs 2.7.x
- this installs `python2` and `pip2`
- `pip2 install hokusai` - worked

### Keys

Need to set all my AWS keys:  `$AWS_ACCESS_KEY_ID, $AWS_SECRET_ACCESS_KEY, $AWS_DEFAULT_REGION and optionally, $AWS_ACCOUNT_ID in your shell / ~/.bash_profile.`

- Found the AWS log in link in potential
- Go here: https://artsy.signin.aws.amazon.com/console
- Then go to users, then find your user (if you don’t have one go to https://github.com/artsy/potential )
- Go to security creds, make an access key
- Migrate those into your shell profile

### Hokusai setup

- need to install docker
- `brew tap caskroom/cask; brew cask install docker`
- `open /Applications/Docker.app`to let it install CLI tools

Need to find all this stuff I’ve never heard of: `--kubectl-version <kubectl version> --s3-bucket <bucket name> --s3-key <file key>`
- Found it in potential: https://github.com/artsy/potential/blob/7eb5716ffcc11315c001f711ea50521a6d254e18/platform/Kubernetes.md

### Start Dev 

- clone PR
- hokusai dev start

This ended up failing:

```sh
~/dev/projects/artsy/js/sites/metaphysics-hokusai master
❯ hokusai dev start
Pulling metaphysics-memcached (memcached:latest)...
latest: Pulling from library/memcached
afeb2bfd31c0: Pull complete
76cfabc68f7b: Pull complete
a25c53f0a67c: Pull complete
0ba510cca659: Pull complete
92ca184d0df4: Pull complete
Digest: sha256:54f000b9f03ddaabe27a5a080d1ad0d39beb5fea80b199664501ac8d5d9cf37a
Status: Downloaded newer image for memcached:latest
Building metaphysics
Step 1/13 : FROM node:8.4.0
8.4.0: Pulling from library/node
aa18ad1a0d33: Pull complete
90f6d19ae388: Pull complete
94273a890192: Pull complete
c9110c904324: Pull complete
788d73c0fb6b: Pull complete
f221bb562f24: Pull complete
14414a6a768d: Pull complete
af6d2b2ad991: Pull complete
Digest: sha256:080488acfe59bae32331ce28373b752f3f848be8b76c2c2d8fdce28205336504
Status: Downloaded newer image for node:8.4.0
 ---> 386940f92d24
Step 2/13 : RUN npm install -g yarn
 ---> Running in b897ccf2d069
npm info it worked if it ends with ok
npm info using npm@5.3.0
npm info using node@v8.4.0
npm http fetch GET 200 https://registry.npmjs.org/yarn 128ms
http fetch GET 200 https://registry.npmjs.org/yarn/-/yarn-1.0.1.tgz 903ms
npm info lifecycle yarn@1.0.1~preinstall: yarn@1.0.1
npm info linkStuff yarn@1.0.1
/usr/local/bin/yarn -> /usr/local/lib/node_modules/yarn/bin/yarn.js
/usr/local/bin/yarnpkg -> /usr/local/lib/node_modules/yarn/bin/yarn.js
npm info lifecycle yarn@1.0.1~install: yarn@1.0.1
npm info lifecycle yarn@1.0.1~postinstall: yarn@1.0.1
+ yarn@1.0.1
added 1 package in 2.062s
npm info ok
 ---> 75a2f776d1d3
Removing intermediate container b897ccf2d069
Step 3/13 : RUN adduser --disabled-password --gecos '' deploy
 ---> Running in 58010b356619
Adding user `deploy' ...
Adding new group `deploy' (1001) ...
Adding new user `deploy' (1001) with group `deploy' ...
Creating home directory `/home/deploy' ...
Copying files from `/etc/skel' ...
 ---> 0781a18580f7
Removing intermediate container 58010b356619
Step 4/13 : RUN mkdir -p /app
 ---> Running in c6fce663d0ee
 ---> 6cce6d591407
Removing intermediate container c6fce663d0ee
Step 5/13 : ADD . /app
 ---> f27b1c831bb9
Removing intermediate container 37e84f2b87c8
Step 6/13 : WORKDIR /app
 ---> d3bc457d563a
Removing intermediate container 226009e7a66d
Step 7/13 : RUN yarn install
 ---> Running in 263c59bf629f
yarn install v1.0.1
[1/5] Validating package.json...
[2/5] Resolving packages...
success Already up-to-date.
Done in 0.68s.
 ---> c888e3003fa6
Removing intermediate container 263c59bf629f
Step 8/13 : RUN chown -R deploy:deploy /app
 ---> Running in d6ce737b39cb
 ---> 745628dfaf06
Removing intermediate container d6ce737b39cb
Step 9/13 : RUN node_modules/.bin/babel index.js config.js -s inline -d build
 ---> Running in a5e506eeb310
index.js -> build/index.js
config.js -> build/config.js
 ---> f53fc210ce31
Removing intermediate container a5e506eeb310
Step 10/13 : RUN node_modules/.bin/babel lib -s inline -d build/lib
 ---> Running in 549b3438327e
lib/all.js -> build/lib/all.js
lib/apis/convection.js -> build/lib/apis/convection.js
lib/apis/delta.js -> build/lib/apis/delta.js

[snip]

build/lib/loaders/loaders_without_authentication/index.js
lib/loggers.js -> build/lib/loggers.js
lib/routing.js -> build/lib/routing.js
lib/throttle.js -> build/lib/throttle.js
lib/timer.js -> build/lib/timer.js
 ---> fb3ef56fd5ee
Removing intermediate container 549b3438327e
Step 11/13 : RUN node_modules/.bin/babel schema -s inline -d build/schema
 ---> Running in adfd68ebd710
schema/aggregations/aggregation_count.js -> build/schema/aggregations/aggregation_count.js
schema/aggregations/filter_artworks_aggregation.js -> build/schema/aggregations/filter_artworks_aggregation.js
schema/aggregations/filter_partners_aggregation.js -> build/schema/aggregations/filter_partners_aggregation.js

[snip]

schema/trending.js -> build/schema/trending.js
schema/types/formatted_number.js -> build/schema/types/formatted_number.js
 ---> b3919600377f
Removing intermediate container adfd68ebd710
Step 12/13 : USER deploy
 ---> Running in 1296b6283468
 ---> cadac06146b3
Removing intermediate container 1296b6283468
Step 13/13 : CMD node build/index.js
 ---> Running in 0569bac2bb7c
 ---> 11642205ee70
Removing intermediate container 0569bac2bb7c
Successfully built 11642205ee70
Successfully tagged hokusai_metaphysics:latest
Creating hokusai_metaphysics-memcached_1 ...
Creating hokusai_metaphysics-memcached_1 ... done
Creating hokusai_metaphysics_1 ...
Creating hokusai_metaphysics_1 ... done
Starting metaphysics-memcached ... error
Starting metaphysics ... error

ERROR: for metaphysics-memcached  Cannot start service metaphysics-memcached: network hokusai_default not found
ERROR: No containers to start
```

**Comment by izakp** [comment]
@orta thanks for the setup notes.  The system Python workaround is a pain.... I want to test on some clean installs.  Do you have a stacktrace from this failure?

Also, yea I excluded the `AWS_*` part from this PR as well as linking to the relevant doc in Potential, as I was just going describing a local setup.  But good find.

I think you ran into this issue with how hokusai leverages docker-compose: https://github.com/docker/compose/issues/3477  Good catch.

Can you try running the command `docker-compose -f ./hokusai/development.yml up` as a workaround?  You just have to run this once... Ctl-C out of it once the stack is running, then `hokusai dev` should work.

**Comment by orta** [comment]
Yep - that worked 

```sh
~/dev/projects/artsy/js/sites/metaphysics-hokusai master 4m 45s
❯ docker-compose -f ./hokusai/development.yml up
Creating network "hokusai_default" with the default driver
Starting hokusai_metaphysics-memcached_1 ...
Starting hokusai_metaphysics-memcached_1 ... done
Starting hokusai_metaphysics_1 ...
Starting hokusai_metaphysics_1 ... done
Attaching to hokusai_metaphysics-memcached_1, hokusai_metaphysics_1
metaphysics_1            | /app/node_modules/passport-oauth2/lib/strategy.js:83
metaphysics_1            |   if (!options.clientID) { throw new TypeError('OAuth2Strategy requires a clientID option'); }
metaphysics_1            |                            ^
metaphysics_1            |
metaphysics_1            | TypeError: OAuth2Strategy requires a clientID option
metaphysics_1            |     at new OAuth2Strategy (/app/node_modules/passport-oauth2/lib/strategy.js:83:34)
metaphysics_1            |     at Object.<anonymous> (/app/build/lib/auth/index.js:29:16)
metaphysics_1            |     at Module._compile (module.js:573:30)
metaphysics_1            |     at Object.Module._extensions..js (module.js:584:10)
metaphysics_1            |     at Module.load (module.js:507:32)
metaphysics_1            |     at tryModuleLoad (module.js:470:12)
metaphysics_1            |     at Function.Module._load (module.js:462:3)
metaphysics_1            |     at Module.require (module.js:517:17)
metaphysics_1            |     at require (internal/module.js:11:18)
metaphysics_1            |     at Object.<anonymous> (/app/build/index.js:63:13)
metaphysics_1            |     at Module._compile (module.js:573:30)
metaphysics_1            |     at Object.Module._extensions..js (module.js:584:10)
metaphysics_1            |     at Module.load (module.js:507:32)
metaphysics_1            |     at tryModuleLoad (module.js:470:12)
metaphysics_1            |     at Function.Module._load (module.js:462:3)
metaphysics_1            |     at Function.Module.runMain (module.js:609:10)
hokusai_metaphysics_1 exited with code 1
^CERROR: Aborting.

~/dev/projects/artsy/js/sites/metaphysics-hokusai master 14s
❯ hokusai dev start
Building metaphysics
Step 1/13 : FROM node:8.4.0
 ---> 386940f92d24
Step 2/13 : RUN npm install -g yarn
 ---> Using cache
 ---> 75a2f776d1d3
Step 3/13 : RUN adduser --disabled-password --gecos '' deploy
 ---> Using cache
 ---> 0781a18580f7
Step 4/13 : RUN mkdir -p /app
 ---> Using cache
 ---> 6cce6d591407
Step 5/13 : ADD . /app
 ---> Using cache
 ---> f27b1c831bb9
Step 6/13 : WORKDIR /app
 ---> Using cache
 ---> d3bc457d563a
Step 7/13 : RUN yarn install
 ---> Using cache
 ---> c888e3003fa6
Step 8/13 : RUN chown -R deploy:deploy /app
 ---> Using cache
 ---> 745628dfaf06
Step 9/13 : RUN node_modules/.bin/babel index.js config.js -s inline -d build
 ---> Using cache
 ---> f53fc210ce31
Step 10/13 : RUN node_modules/.bin/babel lib -s inline -d build/lib
 ---> Using cache
 ---> fb3ef56fd5ee
Step 11/13 : RUN node_modules/.bin/babel schema -s inline -d build/schema
 ---> Using cache
 ---> b3919600377f
Step 12/13 : USER deploy
 ---> Using cache
 ---> cadac06146b3
Step 13/13 : CMD node build/index.js
 ---> Using cache
 ---> 11642205ee70
Successfully built 11642205ee70
Successfully tagged hokusai_metaphysics:latest
hokusai_metaphysics-memcached_1 is up-to-date
Starting metaphysics-memcached ... done
Starting metaphysics ... done

~/dev/projects/artsy/js/sites/metaphysics-hokusai master 17s
```

WRT the original`pip install hokusai` looks like it's also needed `sudo`: 

```
❯ pip install hokusai
Collecting hokusai
  Downloading hokusai-0.3.3-py2-none-any.whl (62kB)
    100% |████████████████████████████████| 71kB 734kB/s
Collecting click==6.6 (from hokusai)
  Using cached click-6.6-py2.py3-none-any.whl
Collecting termcolor==1.1.0 (from hokusai)
Collecting MarkupSafe==0.23 (from hokusai)
Collecting Jinja2==2.8 (from hokusai)
  Using cached Jinja2-2.8-py2.py3-none-any.whl
Collecting boto3==1.4.4 (from hokusai)
  Using cached boto3-1.4.4-py2.py3-none-any.whl
Collecting botocore==1.5.41 (from hokusai)
  Using cached botocore-1.5.41-py2.py3-none-any.whl
Collecting PyYAML==3.12 (from hokusai)
Collecting s3transfer<0.2.0,>=0.1.10 (from boto3==1.4.4->hokusai)
  Using cached s3transfer-0.1.11-py2.py3-none-any.whl
Collecting jmespath<1.0.0,>=0.7.1 (from boto3==1.4.4->hokusai)
  Using cached jmespath-0.9.3-py2.py3-none-any.whl
Collecting docutils>=0.10 (from botocore==1.5.41->hokusai)
  Using cached docutils-0.14-py2-none-any.whl
Collecting python-dateutil<3.0.0,>=2.1 (from botocore==1.5.41->hokusai)
  Using cached python_dateutil-2.6.1-py2.py3-none-any.whl
Collecting futures<4.0.0,>=2.2.0; python_version == "2.6" or python_version == "2.7" (from s3transfer<0.2.0,>=0.1.10->boto3==1.4.4->hokusai)
  Using cached futures-3.1.1-py2-none-any.whl
Collecting six>=1.5 (from python-dateutil<3.0.0,>=2.1->botocore==1.5.41->hokusai)
  Using cached six-1.10.0-py2.py3-none-any.whl
Installing collected packages: click, termcolor, MarkupSafe, Jinja2, futures, jmespath, docutils, six, python-dateutil, botocore, s3transfer, boto3, PyYAML, hokusai


Exception:
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/Library/Python/2.7/site-packages/pip/commands/install.py", line 342, in run
    prefix=options.prefix_path,
  File "/Library/Python/2.7/site-packages/pip/req/req_set.py", line 784, in install
    **kwargs
  File "/Library/Python/2.7/site-packages/pip/req/req_install.py", line 851, in install
    self.move_wheel_files(self.source_dir, root=root, prefix=prefix)
  File "/Library/Python/2.7/site-packages/pip/req/req_install.py", line 1064, in move_wheel_files
    isolated=self.isolated,
  File "/Library/Python/2.7/site-packages/pip/wheel.py", line 345, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/Library/Python/2.7/site-packages/pip/wheel.py", line 316, in clobber
    ensure_dir(destdir)
  File "/Library/Python/2.7/site-packages/pip/utils/__init__.py", line 83, in ensure_dir
    os.makedirs(path)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/Library/Python/2.7/site-packages/click'
```

So in theory you can either `brew install python` and use `pip2` or do a manual installation of pip with sudo, then a sudo pip.

**Comment by izakp** [comment]
Thanks @orta - should be fixed in v0.3.3. You can run `(sudo) pip install --upgrade hokusai` to get the new version, though seems you already worked around the issue.

**Comment by orta** [comment]
👍 I've updated to 0.3.3

I feel like this is not taking my `.env` into account:

```

~/dev/projects/artsy/js/sites/metaphysics-hokusai master ⇣
❯ hokusai dev stop; hokusai dev start
Stopping hokusai_metaphysics-memcached_1 ... done
Building metaphysics
Step 1/13 : FROM node:8.4.0
 ---> 386940f92d24
Step 2/13 : RUN npm install -g yarn
 ---> Using cache
 ---> 75a2f776d1d3
Step 3/13 : RUN adduser --disabled-password --gecos '' deploy
 ---> Using cache
 ---> 0781a18580f7
Step 4/13 : RUN mkdir -p /app
 ---> Using cache
 ---> 6cce6d591407
Step 5/13 : ADD . /app
 ---> f6984b727c2c
Removing intermediate container 7f97d9c9d6b0
Step 6/13 : WORKDIR /app
 ---> d32dfa1a8abb
Removing intermediate container cd333fca3224
Step 7/13 : RUN yarn install
 ---> Running in 5696f0b67124
yarn install v1.0.1
[1/5] Validating package.json...
[2/5] Resolving packages...
success Already up-to-date.
Done in 0.64s.
 ---> 118b6faab66f
Removing intermediate container 5696f0b67124
Step 8/13 : RUN chown -R deploy:deploy /app
 ---> Running in 249f127bd36a
 ---> 9dd4e8737f59
Removing intermediate container 249f127bd36a
Step 9/13 : RUN node_modules/.bin/babel index.js config.js -s inline -d build
 ---> Running in 1817cf95791e
index.js -> build/index.js
config.js -> build/config.js
 ---> b4bc2d487079
Removing intermediate container 1817cf95791e
Step 10/13 : RUN node_modules/.bin/babel lib -s inline -d build/lib
 ---> Running in 663f555e7bc9
lib/all.js -> build/lib/all.js
lib/apis/convection.js -> build/lib/apis/convection.js
...
lib/timer.js -> build/lib/timer.js
 ---> 39af35b97ae8
Removing intermediate container 663f555e7bc9
Step 11/13 : RUN node_modules/.bin/babel schema -s inline -d build/schema
 ---> Running in 1e1f17e418ac
schema/aggregations/aggregation_count.js -> build/schema/aggregations/aggregation_count.js
...
schema/tag.js -> build/schema/tag.js
schema/trending.js -> build/schema/trending.js
schema/types/formatted_number.js -> build/schema/types/formatted_number.js
 ---> 9361a9c77755
Removing intermediate container 1e1f17e418ac
Step 12/13 : USER deploy
 ---> Running in d5547a5ff55e
 ---> be938772e7c9
Removing intermediate container d5547a5ff55e
Step 13/13 : CMD node build/index.js
 ---> Running in 8a4743ce2b6c
 ---> 8f677cd438c7
Removing intermediate container 8a4743ce2b6c
Successfully built 8f677cd438c7
Successfully tagged hokusai_metaphysics:latest
Starting hokusai_metaphysics-memcached_1 ...
Starting hokusai_metaphysics-memcached_1 ... done
Recreating hokusai_metaphysics_1 ...
Recreating hokusai_metaphysics_1 ... done

~/dev/projects/artsy/js/sites/metaphysics-hokusai master ⇣ 1m 15s
❯ hokusai dev logs
Attaching to hokusai_metaphysics_1, hokusai_metaphysics-memcached_1
metaphysics_1            | /app/node_modules/passport-oauth2/lib/strategy.js:83
metaphysics_1            |   if (!options.clientID) { throw new TypeError('OAuth2Strategy requires a clientID option'); }
metaphysics_1            |                            ^
metaphysics_1            |
metaphysics_1            | TypeError: OAuth2Strategy requires a clientID option
metaphysics_1            |     at new OAuth2Strategy (/app/node_modules/passport-oauth2/lib/strategy.js:83:34)
metaphysics_1            |     at Object.<anonymous> (/app/build/lib/auth/index.js:29:16)
metaphysics_1            |     at Module._compile (module.js:573:30)
metaphysics_1            |     at Object.Module._extensions..js (module.js:584:10)
metaphysics_1            |     at Module.load (module.js:507:32)
metaphysics_1            |     at tryModuleLoad (module.js:470:12)
metaphysics_1            |     at Function.Module._load (module.js:462:3)
metaphysics_1            |     at Module.require (module.js:517:17)
metaphysics_1            |     at require (internal/module.js:11:18)
metaphysics_1            |     at Object.<anonymous> (/app/build/index.js:63:13)
metaphysics_1            |     at Module._compile (module.js:573:30)
metaphysics_1            |     at Object.Module._extensions..js (module.js:584:10)
metaphysics_1            |     at Module.load (module.js:507:32)
metaphysics_1            |     at tryModuleLoad (module.js:470:12)
metaphysics_1            |     at Function.Module._load (module.js:462:3)
metaphysics_1            |     at Function.Module.runMain (module.js:609:10)
metaphysics-memcached_1  | Signal handled: Terminated.
metaphysics-memcached_1  | Signal handled: Terminated.
```

But running `yarn start` works fine:

```sh
~/dev/projects/artsy/js/sites/metaphysics-hokusai master ⇣
❯ yarn start
yarn start v1.0.1
$ yarn run dev
[OKAY] Loaded ENV .env File as KEY=VALUE Format
[OKAY] Trimming display Output to 131 Columns
15:57:24 web.1       |  Debugger listening on ws://127.0.0.1:9229/66436cf0-bb38-4925-a590-8fda6fd0b535
15:57:24 web.1       |  For help see https://nodejs.org/en/docs/inspector
15:57:25 web.1       |  {"v":0,"level":30,"name":"newrelic","hostname":"Ortas-MacBook-Pro.local","pid":24124,"time":"2017-09-13T19:57:25.208Z","msg":"Using…
15:57:25 web.1       |  {"v":0,"level":40,"name":"newrelic","hostname":"Ortas-MacBook-Pro.local","pid":24124,"time":"2017-09-13T19:57:25.283Z","msg":"The n…
15:57:25 web.1       |  {"v":0,"level":30,"name":"newrelic","hostname":"Ortas-MacBook-Pro.local","pid":24124,"time":"2017-09-13T19:57:25.286Z","msg":"Could…
15:57:25 web.1       |  {"v":0,"level":30,"name":"newrelic","hostname":"Ortas-MacBook-Pro.local","pid":24124,"time":"2017-09-13T19:57:25.287Z","msg":"Could…
15:57:25 web.1       |  {"v":0,"level":30,"name":"newrelic","hostname":"Ortas-MacBook-Pro.local","pid":24124,"time":"2017-09-13T19:57:25.435Z","msg":"Start…
15:57:26 web.1       |  Debugger attached.
15:57:29 web.1       |  Wed, 13 Sep 2017 19:57:29 GMT express-session deprecated undefined resave option; provide resave option at index.js:109:38
15:57:29 web.1       |  Wed, 13 Sep 2017 19:57:29 GMT express-session deprecated undefined saveUninitialized option; provide saveUninitialized option at in…
15:57:29 web.1       |  Wed, 13 Sep 2017 19:57:29 GMT info Listening on 5001
15:57:31 web.1       |  {"v":0,"level":30,"name":"newrelic","hostname":"Ortas-MacBook-Pro.local","pid":24124,"time":"2017-09-13T19:57:31.210Z","msg":"Repor…
15:57:31 web.1       |  {"v":0,"level":30,"name":"newrelic","hostname":"Ortas-MacBook-Pro.local","pid":24124,"time":"2017-09-13T19:57:31.210Z","msg":"Conne…
```

When I move the `.env` then I get the same issue as the hokusai version:

```sh
~/dev/projects/artsy/js/sites/metaphysics-hokusai master ⇣ 15s
❯ mv .env ._env

~/dev/projects/artsy/js/sites/metaphysics-hokusai master* ⇣
❯ yarn start
yarn start v1.0.1
$ yarn run dev
[WARN] No ENV file found
[OKAY] Trimming display Output to 131 Columns
16:00:00 web.1       |  Debugger listening on ws://127.0.0.1:9229/e2912db7-b0b6-426a-9ea7-4debad3a873c
16:00:00 web.1       |  For help see https://nodejs.org/en/docs/inspector
16:00:03 web.1       |  Debugger attached.
16:00:05 web.1       |  /Users/orta/dev/projects/artsy/js/sites/metaphysics-hokusai/node_modules/passport-oauth2/lib/strategy.js:83
16:00:05 web.1       |    if (!options.clientID) { throw new TypeError('OAuth2Strategy requires a clientID option'); }
16:00:05 web.1       |                                   ^
16:00:05 web.1       |  TypeError: OAuth2Strategy requires a clientID option
16:00:05 web.1       |      at new OAuth2Strategy (/Users/orta/dev/projects/artsy/js/sites/metaphysics-hokusai/node_modules/passport-oauth2/lib/strategy.js…
16:00:05 web.1       |      at Object.<anonymous> (/Users/orta/dev/projects/artsy/js/sites/metaphysics-hokusai/lib/auth/index.js:9:18)
16:00:05 web.1       |      at Module._compile (module.js:573:30)
16:00:05 web.1       |      at babelWatchLoader (/Users/orta/dev/projects/artsy/js/sites/metaphysics-hokusai/node_modules/babel-watch/runner.js:50:13)
16:00:05 web.1       |      at Object.require.extensions.(anonymous function) [as .js] (/Users/orta/dev/projects/artsy/js/sites/metaphysics-hokusai/node_mo…
16:00:05 web.1       |      at Module.load (module.js:507:32)
16:00:05 web.1       |      at tryModuleLoad (module.js:470:12)
16:00:05 web.1       |      at Function.Module._load (module.js:462:3)
16:00:05 web.1       |      at Module.require (module.js:517:17)
16:00:05 web.1       |      at require (internal/module.js:11:18)
16:00:05 web.1       |      at Object.<anonymous> (/Users/orta/dev/projects/artsy/js/sites/metaphysics-hokusai/index.js:16:1)
16:00:05 web.1       |      at Module._compile (module.js:573:30)
16:00:05 web.1       |      at babelWatchLoader (/Users/orta/dev/projects/artsy/js/sites/metaphysics-hokusai/node_modules/babel-watch/runner.js:50:13)
16:00:05 web.1       |      at Object.require.extensions.(anonymous function) [as .js] (/Users/orta/dev/projects/artsy/js/sites/metaphysics-hokusai/node_mo…
16:00:05 web.1       |      at Module.load (module.js:507:32)
16:00:05 web.1       |      at tryModuleLoad (module.js:470:12)
16:00:05 web.1       |  Waiting for the debugger to disconnect...
```

**Comment by izakp** [comment]
@orta pull my latest commits and try again - the Dockerfile "entrypoint" is the `CMD` directive, and  I now have it so that it starts node-foreman, so should pick up your .env file.  This should be a generalized default command and overridden in the yml files in the `./hokusai` folder to customize per-environment.  As such I don't think we need to run this in production with node-foreman launching forever, which launches yet another node process, but did not want to tackle that right now.

File: hokusai/production.yml (hunk 2 above)
Comment (izakp) [reply hokusai/production.yml]: Good call.

**Comment by orta** [comment]
Yep, that works 👍

**Comment by cavvia** [comment]
Great detailed writeup @orta! We definitely need a developer guide beyond the Hokusai documentation, which can deal with this step-by-step for setting up internal projects.

We could also add some of the steps (such as installing a non-prehistoric Python version) as development pre-requisites in our [potential setup script](https://github.com/artsy/potential/blob/master/scripts/setup).

**Comment by cavvia** [comment]
@isac I think we would want a minimum resource request for MP in Production, as it regularly tops out its allocated 2GB on Heroku - could we use a [resource request](https://kubernetes.io/docs/tasks/administer-cluster/memory-default-namespace/#what-if-you-specify-a-containers-request-but-not-its-limit) for that?

**Comment by izakp** [comment]
@cavvia yes we should try and constrain its memory allocation with a request /  limit

**Comment by damassi** [comment]
Tiny comment for users following along: 

> You can then run (in the project root) `hokusai dev` and Metaphysics... 

returns

```sh
metaphysics [izakp-master] % hokusai dev
Usage: hokusai dev [OPTIONS] ACTION

Error: Missing argument "action".
```

Might be good to update PR instructions to include full command so there's no confusion -- `hokusai dev <build|start|etc>`

**Comment by damassi** [comment]
Little DX thing -- it would be nice to see some kind of output once you run `hokusai run dev` that outputs the result, e.g., 

```sh
$ hokusai run dev

Building metaphysics
Step 13/13 : CMD node_modules/.bin/nf start
 ---> Using cache
 ---> 1ff6c029668f
Successfully built 1ff6c029668f
Successfully tagged hokusai_metaphysics:latest
hokusai_metaphysics-memcached_1 is up-to-date
hokusai_metaphysics_1 is up-to-date

[LOG] Server is running on http://localhost:5001. 

Run `hokousai dev stop` to shut down, or `hokusai dev logs --follow` to view logs. 
```

Not sure how generic you can make this, but I had to dig into the config files to learn where to navigate to which is definitely an unconscious deterrent for many users. 

Also, is `--follow` more intuitive than the common `--tail`?

**Comment by damassi** [comment]
(Sorry one more thought! Might be good to sweep over and setup Positron, too; during dev swap I set up `docker-compose` for them but it might be better to unify everything with `hokusai`: https://github.com/artsy/positron/pull/1221.)

**Comment by damassi** [comment]
Going to close this for now until we can coordinate things across the team, but lets re-open this (and other Force PR) when ready

**Pull request closed without merging**